src/Evaluator.hs

```diff
@@ -0,0 +1,214 @@
+module Evaluator where
+
+import           ValueSyntax
+import           Syntax
+import           Environment
+import           Utility
+
+import qualified Data.Text                     as T
+
+evalError :: Show a => Expr a -> String -> b
+evalError expr typeString =
+    error
+        $  "Expression"
+        ++ show expr
+        ++ "\ncannot be evaluated to type: "
+        ++ typeString
+
+exprTypeError :: Show a => Expr a -> String -> b
+exprTypeError expr typeString =
+    error $ "Expression" ++ show expr ++ "\nis not of type " ++ typeString
+
+evalPrimExpr :: PrimExpr -> PrimVal
+evalPrimExpr (IntLit   n) = IntVal n
+evalPrimExpr (CharLit  c) = CharVal c
+evalPrimExpr (FloatLit f) = FloatVal f
+evalPrimExpr (BoolLit  b) = BoolVal b
+evalPrimExpr UnitLit      = UnitVal
+
+evalPrimitive :: Show a => Expr a -> Environment -> IO Value
+evalPrimitive (Primitive primExpr _ _) env =
+    return $ PrimitiveVal $ evalPrimExpr primExpr
+evalPrimitive expr _ = exprTypeError expr "Primitive"
+
+evalVar :: Show a => Expr a -> Environment -> IO Value
+evalVar (Var identifier _ _) env = return $ lookupIdentifier identifier env
+
+evalConstructor :: Show a => Expr a -> Environment -> IO Value
+evalConstructor (Constructor constructorName _ _) env =
+    return $ lookupConstructorName constructorName env
+evalConstructor expr _ = exprTypeError expr "constructor"
+
+evalQualifiedVar :: Show a => Expr a -> Environment -> IO Value
+evalQualifiedVar (QualifiedVar modName identifier _ _) env = 
+    return $ lookupQualifiedVar modName identifier env
+evalQualifiedVar expr _ = exprTypeError expr "QualifiedVar"
+
+evalQualifiedConstructor :: Show a => Expr a -> Environment -> IO Value
+evalQualifiedConstructor (QualifiedConstructor modName constructorName _ _) env= 
+    return $ lookupQualifiedConstructor modName constructorName env
+evalQualifiedConstructor expr _ = exprTypeError expr "QualifiedConstructor"
+
+evalAnd :: Show a => Expr a -> Environment -> IO Value
+evalAnd (And leftExpr rightExpr _ _) env = do
+    leftVal <- eval leftExpr env
+    case leftVal of
+        PrimitiveVal (BoolVal False) -> return leftVal
+        PrimitiveVal (BoolVal True ) -> do
+            rightVal <- eval rightExpr env
+            case rightVal of
+                PrimitiveVal (BoolVal _) -> return rightVal
+                _                        -> evalError rightExpr "bool"
```

Comment:
I think all this dynamic type checking isn't necessary since we could assume that the expressions given as an input are well typed (typechecking should've been done prior to the evaluation).
Then this could be simplified to:
```haskell
evalAnd (And lhs rhs _ _) env = do
  PrimitiveVal (BoolVal lhsVal) <- eval lhs env
  if lhsVal
    then return (PrimitiveVal (BoolVal True))    
    else eval rhs env 
```